### ISSUE TITLE
Add mini-batch/stochastic gradient descent

### ISSUE THREAD

**Issue by KenNewcomb** [KenNewcomb/MLearn#3]
Currently, the only optimizer available is batch gradient descent, where all training examples are used the compute the gradient during each epoch. I'd like to implement stochastic gradient descent, where random samples are used to compute the gradient, and mini-batch gradient descent, where small subsets of the training data (mini-batches) are used to compute the gradient. 

Introduce batch_size parameter, along with a sensible default.

Perhaps these should be in their own class (Optimizer)?
